tensorflow/python/kernel_tests/check_ops_test.py

```diff
@@ -1457,6 +1457,237 @@ def test_raises_when_wrong_type(self):
+
+    raises_static_rank_error(rank_two_shapes, array_ops.ones([1]),
+                             correct_rank=2, actual_rank=1)
+    raises_static_rank_error(rank_three_shapes, array_ops.ones([1, 1]),
+                             correct_rank=3, actual_rank=2)
+    raises_static_rank_error(rank_three_shapes, array_ops.constant(1),
+                             correct_rank=3, actual_rank=0)
+
+  def test_raises_dynamic_incorrect_rank(self):
+    x_value = 5
+    rank_two_shapes = [
+        (1, 1),
+        (1, 3),
+        ('a', 'b'),
+        (None, None)
+    ]
+    with ops.Graph().as_default():
+      x = array_ops.placeholder(dtypes.float32, None)
+
+      for shape in rank_two_shapes:
+        regex = r"Tensor .* must have rank\] \[2\]"
+        self.raises_dynamic_error(shapes={x: shape}, regex=regex,
+                                  feed_dict={x: x_value})
+
+  @test_util.run_in_graph_and_eager_modes
+  def test_correctly_matching(self):
+    u = array_ops.constant(1, name="v")
+    v = array_ops.ones([1, 2], name="v")
+    w = array_ops.ones([3], name="w")
+    x = array_ops.ones([1, 2, 3], name="x")
+    y = array_ops.ones([3, 1, 2], name="y")
+    z = array_ops.ones([2, 3, 1], name="z")
+    assertion = check_ops.assert_shapes({
+        x: ('a', 'b', 'c'),
+        y: ('c', 'a', 'b'),
+        z: ('b', 'c', 'a'),
+        v: ('a', 'b'),
+        w: ('c',),
+        u: 'a'
+    })
+    with ops.control_dependencies([assertion]):
+      out = array_ops.identity(x)
+    self.evaluate(out)
+    assertion = check_ops.assert_shapes({
+        x: (1, 'b', 'c'),
+        y: ('c', 'a', 2),
+        z: ('b', 3, 'a'),
+        v: ('a', 2),
+        w: (3,),
+        u: 1
+    })
+    with ops.control_dependencies([assertion]):
+      out = array_ops.identity(x)
+    self.evaluate(out)
+
+  @test_util.run_in_graph_and_eager_modes
+  def test_variable_length_symbols(self):
+    x = array_ops.ones([4, 1], name="x")
+    y = array_ops.ones([4, 2], name="y")
+    assertion = check_ops.assert_shapes({
+        x: ('num_observations', 'input_dim'),
+        y: ('num_observations', 'output_dim'),
+    })
+    with ops.control_dependencies([assertion]):
+      out = array_ops.identity(x)
+    self.evaluate(out)
+
+  @test_util.run_in_graph_and_eager_modes
+  def test_empty_shapes_dict_no_op(self):
+    assertion = check_ops.assert_shapes({})
+    with ops.control_dependencies([assertion]):
+      out = array_ops.identity(0)
+    self.evaluate(out)
+
+  def raises_static_error(self, shapes, regex):
+    with self.assertRaisesRegexp(ValueError, regex):
+      check_ops.assert_shapes(shapes)
+
+  def raises_dynamic_error(self, shapes, regex, feed_dict):
+    with self.session() as sess:
+      with self.assertRaisesRegexp(errors.InvalidArgumentError, regex):
+        assertion = check_ops.assert_shapes(shapes)
+        with ops.control_dependencies([assertion]):
+          out = array_ops.identity(0)
+        sess.run(out, feed_dict=feed_dict)
+
+
```

Comment:
I was showing your operation to some other core TF devs and someone asked for a unit test that says that you can use any sequence, including a string, to validate shapes. So something like

tf.assert_shapes({
 image: "NCHW",
 kernel: "Cc",
})

would validate a conv kernel. How do you feel about this?

Reply:
I like that idea a lot. 

To enable the same capabilities using e.g. strings we would need a character symbol for `...` and None (*any* size). In the code just pushed it is currently `#` for `...` and  `*` for *any*.

In the code I made it an arbitrary iterable, then not needing the attribute `__len__` but only the `__iter__` and then use the tuple(iterable) constructor to guarantee having `__len__` for the specified shape rank checks.

As certain iterable/sequence types have their own element types, as e.g. `str` having `str` elements, we need to be cast/parse elements to integers for the explicit size check case. To handle it for general/unknown element types the pythonic way I know is a try-catch int parsing attempt. If you prefer another approach I'd be happy to do it differently.

What do you think?

Reply:
I think the string case is good even if it's more limited (no None, number, or ellipsis).

Reply:
I definitely think too the string case is good either way.

Although I think that it would still be neat to have all functionality in the string/general iterable case as well, as it would allow for a compact syntax as a complete alternative. 

If alternative symbols were to be included, perhaps they could instead by the following to have similar meaning as in regular expressions:
*any*: '.'
ellipsis: '*'

Examples:
```python
def matmul(x, y):
  tf.assert_shapes({
      x: (..., 'A', 'B'),
      y: (..., 'B', 'C')
  })
  ...
```
and as there is no need to declare 'A' and 'C' it could be written:	
```python
tf.assert_shapes({
  	x: (..., '.', 'B'),
  	y: (..., 'B', '.')
})
```
or using strings:
```python
tf.assert_shapes({
  x: "*.B",
  y: "*B."
})
```

Reply:
Great. Pushed the symbol changes